File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        n = len(nums)
        left = [nums[0]] * n  
        right = [nums[-1]] * n

        for i in range(1,n):
            left[i] = left[i-1] * nums[i]

        for j in range(n-2, -1, -1):
            right[j] = right[j+1] * nums[j]

        res = [right[1]]

        for i in range(1,n-1):
            res.append(left[i-1] * right[i+1])

        res.append(left[-2])

        return res



        

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} k
     * @return {number}
     */
    kthSmallest(root: TreeNode | null, k: number): number {

        let res;
        let count = k;

        function dfs(node: TreeNode | null) {
            if (!node) {
                return
            }

            dfs(node.left)

            if (count == 0) {
                return
            }

            count--;

            if (count == 0) {
                res = node.val
            }

            dfs(node.right)
        }

        dfs(root);
        return res;
    }
}
